server/state.go

```diff
@@ -255,133 +167,174 @@ func (cs *callState) getClientState(botID, userID string) *CallStateClient {
-		Transcription:          cs.Transcription.getClientState(),
+		HostID:                 cs.GetHostID(),
+		Recording:              getClientStateFromCallJob(cs.Recording),
+		Transcription:          getClientStateFromCallJob(cs.Transcription),
 		DismissedNotification:  dismissed,
 	}
 }
 
 func (cs *callState) getUsersAndStates(botID string) ([]string, []UserStateClient) {
-	users := make([]string, 0, len(cs.Sessions))
-	states := make([]UserStateClient, 0, len(cs.Sessions))
-	for sessionID, state := range cs.Sessions {
+	users := make([]string, 0, len(cs.sessions))
+	states := make([]UserStateClient, 0, len(cs.sessions))
+	for _, session := range cs.sessions {
 		// We don't want to expose to the client that the bot is in a call.
-		if state.UserID == botID {
+		if session.UserID == botID {
 			continue
 		}
-		users = append(users, state.UserID)
-		states = append(states, state.getClientState(sessionID))
+		users = append(users, session.UserID)
+		states = append(states, UserStateClient{
+			SessionID:  session.ID,
+			UserID:     session.UserID,
+			Unmuted:    session.Unmuted,
+			RaisedHand: session.RaisedHand,
+		})
 	}
 	return users, states
 }
 
 func (cs *callState) onlyUserLeft(userID string) bool {
-	for _, state := range cs.Sessions {
-		if state.UserID != userID {
+	var found bool
+	for _, session := range cs.sessions {
+		if session.UserID != userID {
 			return false
 		}
+		found = true
 	}
-	return true
+	return found
 }
 
-func (p *Plugin) kvGetChannelState(channelID string, fromWriter bool) (*channelState, error) {
-	data, appErr := p.KVGet(channelID, fromWriter)
-	if appErr != nil {
-		return nil, fmt.Errorf("KVGet failed: %w", appErr)
+func (p *Plugin) getCallStateFromCall(call *public.Call, fromWriter bool) (*callState, error) {
+	if call == nil {
+		return nil, fmt.Errorf("call should not be nil")
 	}
-	if data == nil {
-		return nil, nil
+
+	state := &callState{
+		Call: *call,
 	}
-	var state *channelState
-	if err := json.Unmarshal(data, &state); err != nil {
-		return nil, err
+
+	participants := make(map[string]struct{}, len(call.Participants))
+	for _, p := range call.Participants {
+		participants[p] = struct{}{}
 	}
-	return state, nil
-}
 
-func (p *Plugin) kvSetChannelState(channelID string, state *channelState) error {
-	p.metrics.IncStoreOp("KVSet")
+	sessions, err := p.store.GetCallSessions(call.ID, db.GetCallSessionOpts{
+		FromWriter: fromWriter,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get call sessions: %w", err)
+	}
+	state.sessions = sessions
 
-	data, err := json.Marshal(state)
+	jobs, err := p.store.GetCallJobs(call.ID, db.GetCallJobOpts{
+		FromWriter: fromWriter,
+	})
 	if err != nil {
-		return fmt.Errorf("failed to marshal channel state: %w", err)
+		return nil, fmt.Errorf("failed to get call jobs: %w", err)
+	}
+	for _, job := range jobs {
+		if job.Type == public.JobTypeRecording && state.Recording == nil {
+			state.Recording = job
```

Comment:
What's the purpose of the `state.Recording == nil` and `state.Transcription == nil` checks?

Reply:
That's a very good question. I meant to leave a comment. The reason is that the returned jobs are sorted by `StartAt` and the `nil` check is because we want to extract only the most recent one because in theory there's no DB constraint so there could be more than one "unfinished". If we were returning a single job type we could `LIMIT` but in this case we want to get all types in one go.